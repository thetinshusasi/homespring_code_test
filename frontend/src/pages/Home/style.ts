import { makeStyles } from '@material-ui/core/styles';
const useStyles = makeStyles({
    root: {
        paddingTop: "2em",
        paddingBottom: "2em",
        paddingLeft: "2.5em",
        paddingRight: "2.5em",
        height: "100vh"
    },
    paginationContainer: {
        width: "100%",
        display: "flex",
        justifyContent: "center"
    },
    searchContainer: {
        display: "flex",
        flexDirection: "column",
        justifyContent: "space-between",
        height: "100vh"
    },
    searchTermPrompt: {
        textAlign: "center",
        marginTop: "2em"
    },
    progress: {
        marginTop: "2em",
        marginBottom: "2em",
        textAlign: "center"
    }

});

export default useStyles

