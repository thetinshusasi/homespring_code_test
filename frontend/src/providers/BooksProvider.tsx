import React, { useEffect, useState } from "react";
import useGetBooksBySearchTerm from "../hooks/useGetBooksBySearchTerm";
import Book from "../models/books/Book"

export interface BooksContextProps {
    books: Book[] | undefined,
    startIndex: number,
    kind: string,
    totalItems: number,
    error: any,
    loading: boolean,
    getBooksBySearchTerm: (val: string) => void
}
interface BooksProviderProps {
    children: React.ReactNode;
}
export const BooksContext = React.createContext<BooksContextProps>({
    books: undefined,
    error: false,
    getBooksBySearchTerm: () => { },
    kind: "",
    loading: false,
    startIndex: 0,
    totalItems: 0

});

const BooksProvider = ({ children }: BooksProviderProps) => {
    const { data, error, getBooksBySearchTerm, loading } = useGetBooksBySearchTerm();
    const { items: books = undefined, startIndex = 0, kind = "", totalItems = 0 } = data || {}

    const [value, setValue] = useState({
        books,
        error,
        getBooksBySearchTerm,
        loading,
        startIndex,
        totalItems,
        kind
    })

    useEffect(() => {
        const { items: books = undefined, startIndex = 0, kind = "", totalItems = 0 } = data || {}
        setValue({
            books,
            error,
            loading,
            startIndex,
            totalItems,
            kind,
            getBooksBySearchTerm

        })


    }, [data, loading, error])


    return (
        <BooksContext.Provider
            value={value}
        >
            {children}
        </BooksContext.Provider>
    );
};

export default BooksProvider;
