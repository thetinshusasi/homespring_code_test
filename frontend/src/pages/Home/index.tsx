import { CircularProgress, Container, Grid, Paper, Typography } from '@material-ui/core';
import { Pagination } from '@material-ui/lab';
import { constants } from 'buffer';
import React, { useCallback, useContext, useEffect, useState } from 'react'
import CardList from '../../components/CardList';
import RatingDropdown from '../../components/RatingDropDown'
import Search from '../../components/Search';
import Book from '../../models/books/Book';
import { BooksContext } from '../../providers/BooksProvider';
import literals from '../../utils/literals';
import useStyles from './style';
import { RATINGS } from '../../utils/constants';


export interface HomeProps {
    ratings: number[],
}

export default function Home() {
    const classes = useStyles();
    const [searchTerm, setSearchTerm] = useState("");
    const { books, error, getBooksBySearchTerm, kind, loading, startIndex, totalItems } = useContext(BooksContext);
    const [currentBooks, setCurrentBooks] = useState<Book[] | undefined>(undefined)
    const [listSlice, setListSlice] = useState<Book[] | undefined>(undefined)

    const [currentBooksCount, setCurrentBooksCount] = useState<number>(0)
    const [currentPageCount, setCurrentPageCount] = useState<number>(0)
    const [filterRating, setFilterRating] = useState<number | "">("")



    const onChange = (val: string) => {
        setSearchTerm(val)
    }

    useEffect(() => {

        setFilterRating("")
        setCurrentBooks(undefined)
        setCurrentBooksCount(0)
        setCurrentPageCount(0)
        setListSlice(undefined)


        getBooksBySearchTerm(searchTerm)
    }, [searchTerm])

    const onRatingsChange = (rating: number | "") => {

        if (!currentBooksCount) {
            setFilterRating("")
            return
        }
        setFilterRating(rating)
    }


    const onPageChange = (evt: any, page: number) => {

        const totalCount = ((page - 1) * 10) + 10
        if (totalCount < currentBooksCount) {
            setListSlice(currentBooks?.slice((page - 1) * 10, totalCount))
            return
        }
        setListSlice(currentBooks?.slice((page - 1) * 10))


    }


    useEffect(() => {
        setCurrentBooks(books)
        setCurrentBooksCount(books?.length || 0)
        setCurrentPageCount(books?.length ? Math.floor(books.length * 0.1) : 0)

        const bookLen = books?.length || 0
        if (!bookLen) {

            setListSlice(undefined)
            return
        }
        if (bookLen && bookLen < 10) {

            setListSlice(books?.slice(0))
            return
        }
        if (bookLen && bookLen >= 10) {

            setListSlice(books?.slice(0, 10))
            return
        }

    }, [loading, error, books])

    useEffect(() => {
        let bookLen
        if (filterRating) {
            const tempBooks = books?.filter(item => item.volumeInfo?.averageRating && item.volumeInfo?.averageRating >= filterRating)
            setCurrentBooks(tempBooks)
            setCurrentBooksCount(tempBooks?.length || 0)
            setCurrentPageCount(tempBooks?.length ? Math.floor(tempBooks.length * 0.1) : 0)
            bookLen = tempBooks?.length || 0
            if (!bookLen) {

                setListSlice(undefined)
                return
            }
            if (bookLen && bookLen < 10) {

                setListSlice(tempBooks?.slice(0))
                return
            }
            if (bookLen && bookLen >= 10) {

                setListSlice(tempBooks?.slice(0, 10))
                return
            }
            return
        }
        setCurrentBooks(books)
        setCurrentBooksCount(books?.length || 0)
        setCurrentPageCount(books?.length ? Math.floor(books.length * 0.1) : 0)
        bookLen = books?.length || 0
        if (!bookLen) {

            setListSlice(undefined)
            return
        }
        if (bookLen && bookLen < 10) {

            setListSlice(books?.slice(0))
            return
        }
        if (bookLen && bookLen >= 10) {

            setListSlice(books?.slice(0, 10))
            return
        }
        return

    }, [filterRating])



    return (
        <React.Fragment>
            <Container disableGutters maxWidth={false} className={classes.root}>
                <Grid container spacing={10} >
                    <Grid item md={3} xs={12} >
                        <RatingDropdown ratings={RATINGS} onSelectChange={onRatingsChange} values={filterRating} />
                    </Grid>
                    <Grid item md={8} xs={12} className={classes.searchContainer}>
                        <div>
                            <Search onChange={onChange} value={searchTerm} />
                            {loading && <div className={classes.progress}>
                                <CircularProgress />
                            </div>}
                            {searchTerm && !loading && !error ? <CardList books={listSlice} /> : null}
                            {!searchTerm && !loading && !error ? <Typography variant="h6" gutterBottom className={classes.searchTermPrompt}>
                                {literals.searchPromptMessage}
                            </Typography> : null}

                        </div>
                        {!loading && !error && currentPageCount ? <div className={classes.paginationContainer}>
                            <Pagination count={currentPageCount} variant="outlined" shape="rounded" onChange={onPageChange} />
                        </div> : null}
                    </Grid>
                </Grid>
            </Container>
        </React.Fragment >)
}
