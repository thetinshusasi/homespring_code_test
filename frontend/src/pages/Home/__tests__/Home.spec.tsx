
import React from 'react';
import Home from '../index';
import BooksProvider, { BooksContextProps } from '../../../providers/BooksProvider';
import { fireEvent, getAllByText, render } from '@testing-library/react';

// has to be defined like this cannot be global

describe('Home Component', () => {
    let props;

    beforeEach(() => {
        props = {};
    });

    it('>> renders component', () => {

        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: () => { },
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };
        const { container } = render(
            <BooksProvider value={contextValue}>
                <Home />
            </BooksProvider>
        );
        expect(container).toMatchSnapshot();
    });

    it('>> renders component and contains search component', async () => {
        const mockFn = jest.fn()
        const contextValue: BooksContextProps = {
            books: undefined,
            error: false,
            getBooksBySearchTerm: mockFn,
            kind: "",
            loading: false,
            startIndex: 0,
            totalItems: 0
        };
        const { container, getByText } = render(
            <BooksProvider value={contextValue}>
                <Home />
            </BooksProvider>
        );

        expect(getByText('Filter your results'))


    });


});
