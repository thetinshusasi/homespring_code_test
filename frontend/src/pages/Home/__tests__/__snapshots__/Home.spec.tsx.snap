// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component >> renders component 1`] = `
<div>
  <div
    class="MuiContainer-root makeStyles-root-1 MuiContainer-disableGutters"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-10"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiTypography-root makeStyles-heading-8 MuiTypography-h6 MuiTypography-gutterBottom"
            >
              Filter your results
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div>
              <div
                class="MuiFormControl-root makeStyles-formControl-6"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink="false"
                  id="demo-simple-select-outlined-label"
                >
                  All books
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                  data-testid="ratingSelect"
                >
                  <div
                    aria-haspopup="listbox"
                    aria-labelledby="demo-simple-select-outlined-label demo-simple-select-outlined"
                    class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
                    id="demo-simple-select-outlined"
                    role="button"
                    tabindex="0"
                  >
                    <span>
                      â€‹
                    </span>
                  </div>
                  <input
                    aria-hidden="true"
                    class="MuiSelect-nativeInput"
                    tabindex="-1"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-11"
                    >
                      <span>
                        All books
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root makeStyles-searchContainer-3 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8"
      >
        <div>
          <div
            class="MuiFormControl-root makeStyles-margin-13"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
              data-testid="searchInp"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  scale="3"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
                id="input-with-icon-adornment"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="makeStyles-progress-5"
          >
            <div
              class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
